pkg/adsc/delta.go

```diff
@@ -0,0 +1,736 @@
+//     HandlerContext.RegisterDependency.
+//
+// An example of a handler registration is as follows:
+//
+//	clusterHandler := Register(func(ctx HandlerContext, res *cluster.Cluster, event Event) {
+//	  if event == EventDelete {
+//	    return
+//	  }
+//	  ctx.RegisterDependency(v3.SecretType, ExtractClusterSecretResources(t, res)...)
+//	  ctx.RegisterDependency(v3.EndpointType, ExtractEdsClusterNames([]*cluster.Cluster{res})...)
+//	})
+//
+// It means that when a cluster is added or updated, the client will trigger requests for the
+// secrets and endpoints that the cluster depends on.
+//
+// An example of register handlers:
+//
+//	handlers := []Option{
+//	  clusterHandler,
+//	  Watch[*cluster.Cluster]("*"),
+//	  listenerHandler,
+//	  Watch[*listener.Listener]("*"),
+//	  endpointsHandler,
+//	  routesHandler,
+//	  secretsHandler,
+//	}
+//
+// client := NewDelta("localhost:8080", handlers...)
+//
+// It means that the client will watch all clusters and listeners, and trigger resource events for
+// clusters, listeners, endpoints, routes and secrets that the clusters and listeners depend on.
+type Client struct {
+	cfg      *DeltaADSConfig
+	handlers map[string]HandlerFunc
+	// tree is a map where each key is a `resourceKey` (comprising the resource name and typeURL)
+	// and each added resource is a `resourceNode`. This tree structure represents the dynamic state
+	// and relationships of resources.
+	tree map[resourceKey]resourceNode
+
+	xdsClient discovery.AggregatedDiscoveryService_DeltaAggregatedResourcesClient
+	conn      *grpc.ClientConn
+
+	// initialWatches is the list of resources we are watching on startup
+	initialWatches []resourceKey
+
+	// sendNodeMeta is set to true if the connection is new - and we need to send node meta
+	sendNodeMeta atomic.Bool
+
+	mutex sync.RWMutex
+	// lastReceived message, by type
+	lastReceived    map[string]*discovery.DeltaDiscoveryResponse
+	deltaXDSUpdates chan *discovery.DeltaDiscoveryResponse
+
+	// errChan is used to signal errors from the delta stream
+	errChan chan error
+
+	// closed is set to true when the client is closed
+	closed bool
+}
+
+func (c *Client) trigger(ctx *handlerContext, typeURL string, r *discovery.Resource, event Event) error {
+	res := newProto(typeURL)
+	if r != nil && res != nil {
+		if err := r.Resource.UnmarshalTo(res); err != nil {
+			return err
+		}
+	}
+	handler, f := c.handlers[typeURL]
+	if !f {
+		deltaLog.Warnf("ignoring unknown type %v", typeURL)
+		return nil
+	}
+	handler(ctx, res, event)
+	return nil
+}
+
+// getProtoMessageType returns the Golang type of the proto with the specified name.
+func newProto(tt string) proto.Message {
+	name := protoreflect.FullName(strings.TrimPrefix(tt, resource.APITypePrefix))
+	t, err := protoregistry.GlobalTypes.FindMessageByName(name)
+	if err != nil || t == nil {
+		return nil
+	}
+	return t.New().Interface()
+}
+
+func (c *Client) Run(ctx context.Context) error {
```

Comment:
Hello, is this feature currently enabled? I don't see this method called anywhere.